### PR TITLE
Ensure surface normals and wireframes are using Models internally

### PR DESCRIPTION
**Pull request by brisvag**
# Description
Make sure that surface normals are always using the `SurfaceNormal` model internally (now they can sometimes be dict, which breaks a few assumptions in other parts of the codebase). See napari/napari-animation#145.

I also took the chance to add an event for them, and do the same fix for the wireframe.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Review by alisterburt — APPROVED** [review]
setters look good, thanks! tests are a bit vague so maybe a docstring with the expected behaviour detailed (dict is accepted, model should always be `SurfaceNormals`) would be good - non blocking though

File: napari/layers/surface/surface.py [hunk 1]
```diff
@@ -238,8 +240,11 @@ def __init__(
         # Shading mode
         self._shading = shading
 
-        self.wireframe = wireframe or SurfaceWireframe()
-        self.normals = normals or SurfaceNormals()
+        self._wireframe = SurfaceWireframe()
+        self._normals = SurfaceNormals()
+
+        self.wireframe = wireframe
+        self.normals = normals
```
Comment (Czaki) [inline napari/layers/surface/surface.py]: why is this change required?

File: napari/layers/surface/surface.py [hunk 2]
```diff
@@ -355,6 +360,42 @@ def shading(self, shading):
...
+    def wireframe(self, wireframe):
+        if wireframe is None:
+            self._wireframe.reset()
+        elif isinstance(wireframe, SurfaceWireframe):
+            self._wireframe = wireframe
+        elif isinstance(wireframe, dict):
+            self._wireframe = SurfaceWireframe(**wireframe)
+        else:
+            raise ValueError(
+                f'wireframe should be a dict or SurfaceWireframe, got {type(wireframe)}'
```
Comment (Czaki) [inline napari/layers/surface/surface.py]: Why you do not point here that `None` is acceptable?

File: napari/layers/surface/surface.py [hunk 3]
```diff
@@ -355,6 +360,42 @@ def shading(self, shading):
             self._shading = Shading(shading)
         self.events.shading(value=self._shading)
 
+    @property
+    def wireframe(self):
```
Comment (Czaki) [inline napari/layers/surface/surface.py]:
```suggestion
    def wireframe(self) -> SurfaceWireframe:
```
for better IDE support

File: napari/layers/surface/surface.py [hunk 4]
```diff
@@ -355,6 +360,42 @@ def shading(self, shading):
...
+        elif isinstance(wireframe, dict):
+            self._wireframe = SurfaceWireframe(**wireframe)
+        else:
+            raise ValueError(
+                f'wireframe should be a dict or SurfaceWireframe, got {type(wireframe)}'
+            )
+        self.events.wireframe(value=self._wireframe)
+
+    @property
+    def normals(self):
```
Comment (Czaki) [inline napari/layers/surface/surface.py]:
```suggestion
    def normals(self) -> SurfaceNormals:
```

as above

File: napari/layers/surface/surface.py [hunk 5]
```diff
@@ -355,6 +360,42 @@ def shading(self, shading):
             self._shading = Shading(shading)
         self.events.shading(value=self._shading)
 
+    @property
+    def wireframe(self):
+        return self._wireframe
+
+    @wireframe.setter
+    def wireframe(self, wireframe):
```
Comment (Czaki) [inline napari/layers/surface/surface.py]:
```suggestion
    def wireframe(self, wireframe: Union[dict, SurfaceWireframe, None]):
```

File: napari/layers/surface/surface.py [hunk 6]
```diff
@@ -355,6 +360,42 @@ def shading(self, shading):
...
+                f'wireframe should be a dict or SurfaceWireframe, got {type(wireframe)}'
+            )
+        self.events.wireframe(value=self._wireframe)
+
+    @property
+    def normals(self):
+        return self._normals
+
+    @normals.setter
+    def normals(self, normals):
```
Comment (Czaki) [inline napari/layers/surface/surface.py]:
```suggestion
    def normals(self, normals: Union[dict, SurfaceNormals, None]):
```

File: napari/layers/surface/surface.py (hunk 1 above)
Comment (brisvag) [reply napari/layers/surface/surface.py]: Cause if `None` is passed to the setter, I'm running `self._normals.reset()`, so we need to have a model initialized.

File: napari/layers/surface/surface.py (hunk 2 above)
Comment (brisvag) [reply napari/layers/surface/surface.py]: oops!

**Review by Czaki — COMMENTED** [review]
Maybe also nice to add type annotation

File: napari/layers/surface/surface.py (hunk 1 above)
Comment (Czaki) [reply napari/layers/surface/surface.py]: My bad. I do not notice `_` at the begin of the variable name.

**Comment by brisvag** [comment]
Actually, I'm not super happy with the current setters, cause what I really wanted to do was also run `model.update(**dict(new_value))` for the cases of `dict` and `SurfaceNormals`, which would ensure that we update inplace so we don't mess up event sources etc...

Unfortunately, this fails because anything that has `allow_mutation=False` (several things in `SurfaceNormals` model) straight up fails if it's present in the dict. I could `pop` it, but it seemed convoluted... but maybe that's the better solution

**Comment by brisvag** [comment]
Ok, I tried to implement it the "proper way". This also allows to add a connection of the `normals` event to the layer events themselves (mirroring the implementation of `Image.events.plane`).

**Review by kephale — APPROVED** [review]
This looks great. I also just ran into the napari-animation issue: https://github.com/napari/napari-animation/issues/145 because I had the same error. This looks like it should catch all of the cases that was letting `dict`s float around as attribute values.

[edit]
I'll also note that I am using this patch to work around the napari-animation issue. Now surface animations work nicely.

**Comment by brisvag** [comment]
Thanks for the ping, I forgot about this :P Merging!
